src/librbd/ImageWatcher.cc

```diff
@@ -628,6 +628,14 @@ void ImageWatcher<I>::handle_request_lock(int r) {
     return;
   }
 
+  // kick-start ExclusiveLock state machine to detect client blocklisting
+  if (m_watch_blocklisted) {
+    lderr(m_image_ctx.cct) << this << " make lock detect client blocklisting"
+                           << dendl;
+    m_image_ctx.exclusive_lock->handle_peer_notification(0);
```

Comment:
> I would move this to `schedule_request_lock` -- it's already testing whether the watch is registered, so `} else if (m_watch_blocklisted) { ` would fit nicely there.
> 

I like this idea.

> Or actually, is there a reason this can't be done directly `handle_rewatch_complete`?

Yes, we can do that. I wanted to kickstart the Exclusive lock state machine in the code path specific to handling request lock notification, and didn't want to do it in the  `handle_rewatch_complete()` that is called as part of the general  watcher's rewatch mechanism.